Hw03 arraylist

Bryantsev opened the pull request:
ДЗ 3

Bryantsev committed:
Добавлен класс HelloOtus

Bryantsev committed:
Путь, содержащий "target", включен в .gitignore

Bryantsev committed:
Рефакторинг для повышения читабельности кода

Bryantsev committed:
Удален умолчательный код

Bryantsev committed:
Первый релиз DIYarrayList

Bryantsev committed:
В DIYarrayList выделен размер массива отдельно от емкости по аналогии с ArrayList. В итераторе добавлен учет последнего возвращенного элемента.

Bryantsev committed:
Исправлен Main класс

stvort commented on hw03-arraylist/pom.xml:
```diff
@@ -0,0 +1,44 @@
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>hw03-arraylist</artifactId>
```
Разве 3?

Bryantsev replied:
По номеру урока, а не порядковый )

stvort commented on hw01-maven/src/main/java/ru/otus/lesson1/HelloOtus.java:
```diff
@@ -0,0 +1,63 @@
+package ru.otus.lesson1;
```
Лучше чтобы в ПР попадали коммиты только от текущей работы

Bryantsev replied:
Исправлюсь

stvort commented on hw03-arraylist/src/main/java/ru/otus/lesson3/DIYarrayList.java:
```diff
@@ -0,0 +1,244 @@
+
+    @Override
+    public boolean add(E t) {
+        checkAdequacyCapacity(size + 1);
```
Рост массива на 1 при добавлении каждого нового элемента превращает add в очень дорогую операцию. Лучше расти с запасом, но выставляя наружу только реально добавленные элементы

Bryantsev replied:
Изменил. Это уже совсем копия ArrayList получается, поэтому и не добавил сначала )

stvort commented on hw03-arraylist/src/main/java/ru/otus/lesson3/DIYarrayList.java:
```diff
@@ -0,0 +1,244 @@
+
+    @Override
+    public Object[] toArray() {
+        return dataList;
```
1. А если Capacity > size
2. Возвращается ссылка на внутренний массив. Т.о. можно будет менять его извне

Bryantsev replied:
Исправил

stvort commented on hw03-arraylist/src/main/java/ru/otus/lesson3/DIYarrayList.java:
```diff
@@ -0,0 +1,244 @@
+
+    @Override
+    public E get(int index) {
+        return (E) dataList[index];
```
А тут `checkRange(index);`?

Bryantsev replied:
Добавил

stvort commented on hw03-arraylist/src/main/java/ru/otus/lesson3/DIYarrayList.java:
```diff
@@ -0,0 +1,244 @@
+
+    @Override
+    public String toString() {
+        return Arrays.toString(dataList);
```
А если capacity > size?

Bryantsev replied:
Исправил

stvort commented on hw03-arraylist/src/main/java/ru/otus/lesson3/Main.java:
```diff
@@ -0,0 +1,34 @@
+        System.out.println("Sorted list1: " + list1);
+
+        Collections.sort(list1, Comparator.reverseOrder());
+        System.out.println("Reverse sorted list1: " + list1);
```
А `Collections.addAll`?

Bryantsev replied:
Добавил

Bryantsev committed:
Исправлены замечания

1) Добавлена проверка Collections.addAll();
2) Добавлены недостающие проверки и параметризированная функция toArray;
3) toArray() теперь возвращает копию массива с учетом реально заполненных элементов;
4) Добавлен шаг для увеличения емкости массива (по умолчанию 10) и конструктор, в котором можно его задать вместе с начальной емкостью
5) Исправлен toString

stvort commented on hw03-arraylist/src/main/java/ru/otus/lesson3/DIYarrayList.java:
```diff
@@ -0,0 +1,256 @@
+ */
+public class DIYarrayList<E> implements List<E> {
+
+    private Object[] dataList;
```
А что если сразу сделать `E[] dataList`? А положить в него `Object[]`

Bryantsev replied:
Так делать нельзя - компилятор ругается )
Из описания генериков: все генерик-типы удаляются при компиляции, из-за чего тип массива будет неизвестен в runtime, а массив должен знать свой тип, чтобы проверять совместимость присваиваемых элементов.

stvort commented on hw03-arraylist/src/main/java/ru/otus/lesson3/DIYarrayList.java:
```diff
@@ -0,0 +1,256 @@
+        System.arraycopy(dataList, 0, a, 0, size);
+        if (a.length > size)
+            a[size] = null;
+        return a;
```
Предлагаю обратить внимание на форматирование кода. Лучше длиннее, но читабельнее, чем короче и непонятно)

Bryantsev replied:
Не понял насчет форматирования - я использую такое же как у меня на работе принято руководством )
Что в выделенном куске кода можно сделать длиннее и читабельнее?
